### DISCUSSION TITLE
Querying style information in .slint

### DISCUSSION THREAD

**Discussion by Blquinn** [slint-ui/slint#3685] [General]
I was wondering if there is any way to get variables from the currently applied style in slint. 

For example can I use the style's background color in my own slint files? It would be difficult to make the theme consistent without such a mechanism. I can't find a reference to anything like that in the docs.

I was also wondering what the process for creating a custom style, potentially one that is based on another style, like material, would look like. Is such a thing possible at the moment?

Thanks!

**Comment by ghost** [comment]
Hi @Blquinn,

short answer: yes it is possible but there is no build in mechanism to access the resources yet.

Long answer: for each style you can find a `styling.slint` file (except the qt style). These files providing the styling resource like `brushes` and `Typography` that is relevant for the corresponding style. If you want to access these resource e.g. to create a custom widget that is consistent to that style you have to copy that file in your own library. 

If you looking e.g. for the material styleing file you can check in the repo: `internal/compiler/widgets/material-base/styling.slint`

We want to make theses resources per style accessible trough a public api in the future.

**Reply by Blquinn** [reply]
Would you say this is the recommended way to extend the existing widgets at the moment, to copy paste these files as needed into my project?

I've used QML in the past and their controls typically will have some properties to customize them like here https://doc.qt.io/qt-6/qtquickcontrols-customize.html#customizing-a-control

Is there any chance slint might do something similar in the future?

**Comment by ghost** [comment]
I've created an issue for your request: https://github.com/slint-ui/slint/issues/3687

**Reply by Blquinn** [reply]
Thank you!

**Comment by ogoffart** [comment]
The undocumented way to do it right now is to use StyleMetrics 

```
import {StyleMEtrics} from "std-widgets.slint";
```

And the following variables are available on it:

https://github.com/slint-ui/slint/blob/d9672c5a217dc218ba7aef4de30168f4f8ecaa2e/internal/compiler/widgets/fluent-base/std-widgets-impl.slint#L14-L25

This is not documented, but we even use it in our examples, so i guess this is public interface by now.
But we really need  better way to expose the palette

**Reply by Blquinn** [reply]
This is definitely a nice start, but I really would need access to the entire Palette in my application to make the style truly consistent. For example what I am trying to accomplish to to create a square button, which should look exactly like the style's normal button, except for the fact that it is square. In order to do this I need to know the button's background-color, highlight-color, border-colors, accessible-border-color, on-background-color etc. So it's quite a lot of properties that I would need.

I understand that each style has it's own unique palette, so it might be difficult/impossible to create a unified palette object.
